edgeless_orc/src/orchestrator.rs

```diff
@@ -145,6 +146,133 @@ impl Orchestrator {
+                                Some(UpdatePeersRequest::Add(node_id, invocation_url))
+                            }
+                        }
+                        UpdateNodeRequest::Deregistration(node_id) => {
+                            if let None = clients.get(&node_id) {
+                                // There is no client with that node_id
+                                None
+                            } else {
+                                clients.remove(&node_id);
+                                Some(UpdatePeersRequest::Del(node_id))
+                            }
+                        }
+                    };
+
+                    // If no operation was done (either a new node was already
+                    // present with same agent/invocation URLs or a deregistering
+                    // node did not exist) we accept the command.
+                    let mut response = edgeless_api::function_instance::UpdateNodeResponse::Accepted;
+
+                    if let Some(msg) = msg {
+                        // Update the orchestration logic with the new set of nodes.
+                        orchestration_logic.update_nodes(clients.keys().cloned().collect());
+
+                        // Update all the peers (including the node, unless it
+                        // was a deregister operation).
+                        let mut num_failures: u32 = 0;
+                        for (_node_id, client) in clients.iter_mut() {
+                            if let Err(_) = client.api.function_instance_api().update_peers(msg.clone()).await {
+                                num_failures += 1;
+                            }
+                        }
+
+                        // Only with registration, we also update the new node
+                        // by adding as peers all the existing nodes.
+                        if let Some(this_node_id) = this_node_id {
+                            let mut new_node_client = clients.get_mut(&this_node_id).unwrap().api.function_instance_api();
+                            for (other_node_id, client_desc) in clients.iter_mut() {
+                                if other_node_id.eq(&this_node_id) {
+                                    continue;
+                                }
+                                if let Err(_) = new_node_client
+                                    .update_peers(UpdatePeersRequest::Add(*other_node_id, client_desc.invocation_url.clone()))
+                                    .await
+                                {
+                                    num_failures += 1;
+                                }
+                            }
+                        }
+
+                        response = match num_failures {
+                            0 => edgeless_api::function_instance::UpdateNodeResponse::Accepted,
+                            _ => edgeless_api::function_instance::UpdateNodeResponse::ResponseError(edgeless_api::common::ResponseError {
```

Comment:
Not sure if that should really return a failure. This forces the system to be tightly coupled in a synchronous way.
When adding a node, I would assume that my return value is not affected by updates to other nodes.

Reply:
For now, since the orchestrator does not take specific actions upon failure, I think it is best to return failure if even a single node fails.

Later on, when the orchestrator will manage failures, then the situation will be different.

Example: assume there is one node that does not reply correctly.

- Right now its misbehaviour can affect the workflows deployed in the domain, so it is safer to return failure to let the new node that _something_ is wrong.
- In the future, the orchestrator can isolate the misbehaving node straight away (this requires updating the peers in the domain, "migrating" its function instances if needed, etc) and only after that it is possible to return an Accepted message.